benchmarks/benchmark_utils.py

```diff
@@ -0,0 +1,297 @@
+        total_users = total_users[1:]
+        test_users = test[USER_COL].unique()
+        test_users = np.intersect1d(test_users, total_users)
+        users_items = cartesian_product(test_users, total_items)
+        users_items = pd.DataFrame(users_items, columns=[USER_COL, ITEM_COL])
+        training_removed = pd.merge(users_items, train.astype(str), on=[USER_COL, ITEM_COL], how='left')
+        training_removed = training_removed[training_removed[RATING_COL].isna()][[USER_COL, ITEM_COL]]
+        top_k_scores = score(model, 
+                             test_df=training_removed,
+                             user_col=USER_COL, 
+                             item_col=ITEM_COL, 
+                             prediction_col=PREDICTION_COL, 
+                             top_k=TOP_K)
+    return top_k_scores, t
+
+
+def prepare_training_ncf(train):
+    data = NCFDataset(train=train, 
+                      col_user=USER_COL,
+                      col_item=ITEM_COL,
+                      col_rating=RATING_COL,
+                      col_timestamp=TIMESTAMP_COL,
+                      seed=SEED)
+    return data
+
+
+def train_ncf(params, data):
+    model = NCF(n_users=data.n_users, n_items=data.n_items, **params)
+    with Timer() as t:
+        model.fit(data)
+    return model, t
+
+
+def recommend_k_ncf(model, test, train):
+    with Timer() as t: 
+        users, items, preds = [], [], []
+        item = list(train[ITEM_COL].unique())
+        for user in train[USER_COL].unique():
+            user = [user] * len(item) 
+            users.extend(user)
+            items.extend(item)
+            preds.extend(list(model.predict(user, item, is_list=True)))
+        top_k_scores = pd.DataFrame(data={USER_COL: users, ITEM_COL:items, PREDICTION_COL:preds})
+        merged = pd.merge(train, top_k_scores, on=[USER_COL, ITEM_COL], how="outer")
+        top_k_scores = merged[merged[RATING_COL].isnull()].drop(RATING_COL, axis=1)
+    return top_k_scores, t
+
+
+def train_sar(params, data):
+    model = SARSingleNode(**params)
+    model.set_index(data)    
+    with Timer() as t:
+        model.fit(data)
+    return model, t
+    
+
+def recommend_k_sar(model, test, train):
+    with Timer() as t:
+        top_k_scores = model.recommend_k_items(test)
+    return top_k_scores, t
+
+
+def rating_metrics_pyspark(test, predictions):
+    rating_eval = SparkRatingEvaluation(test, 
+                                        predictions, 
+                                        col_user=USER_COL, 
+                                        col_item=ITEM_COL, 
+                                        col_rating=RATING_COL, 
+                                        col_prediction=PREDICTION_COL)
+    return {
+        "RMSE": rating_eval.rmse(),
+        "MAE": rating_eval.mae(),
+        "R2": rating_eval.exp_var(),
+        "Explained Variance": rating_eval.rsquared()
+    }
+    
+    
+def ranking_metrics_pyspark(test, predictions, k=10):

```

Comment:
Can we just directly call the method?

Reply:
Hey guys, I've been looking at this. I think it would be too much to add all the metrics code here, right now this is the code in functions:
```
def rating_metrics_pyspark(test, predictions):
    rating_eval = SparkRatingEvaluation(test, 
                                        predictions, 
                                        col_user=USER_COL, 
                                        col_item=ITEM_COL, 
                                        col_rating=RATING_COL, 
                                        col_prediction=PREDICTION_COL)
    return {
        "RMSE": rating_eval.rmse(),
        "MAE": rating_eval.mae(),
        "R2": rating_eval.exp_var(),
        "Explained Variance": rating_eval.rsquared()
    }
    
    
def ranking_metrics_pyspark(test, predictions, k=10):
    rank_eval = SparkRankingEvaluation(test, 
                                       predictions, 
                                       k=k, 
                                       col_user=USER_COL, 
                                       col_item=ITEM_COL, 
                                       col_rating=RATING_COL, 
                                       col_prediction=PREDICTION_COL, 
                                       relevancy_method="top_k")
    return {
        "MAP": rank_eval.map_at_k(),
        "nDCG@k": rank_eval.ndcg_at_k(),
        "Precision@k": rank_eval.precision_at_k(),
        "Recall@k": rank_eval.recall_at_k()
    }
    
    
def rating_metrics_python(test, predictions):
    cols = {
        "col_user": USER_COL, 
        "col_item": ITEM_COL, 
        "col_rating": RATING_COL, 
        "col_prediction": PREDICTION_COL
    }
    return {
        "RMSE": rmse(test, predictions, **cols),
        "MAE": mae(test, predictions, **cols),
        "R2": rsquared(test, predictions, **cols),
        "Explained Variance": exp_var(test, predictions, **cols)
    }
    
    
def ranking_metrics_python(test, predictions, k=10):
    cols = {
        "col_user": USER_COL, 
        "col_item": ITEM_COL, 
        "col_rating": RATING_COL, 
        "col_prediction": PREDICTION_COL
    }
    return {
        "MAP": map_at_k(test, predictions, k=k, **cols),
        "nDCG@k": ndcg_at_k(test, predictions, k=k, **cols),
        "Precision@k": precision_at_k(test, predictions, k=k, **cols),
        "Recall@k": recall_at_k(test, predictions, k=k, **cols)
    }
```
then apart from that we need to select which algos are spark and which cpu. If we add all these code in the loop is going to be very big. What do you guys think?

Reply:
Are you asking about the size of the code or about computational time? The latter should be faster now with the LRU cache.

Reply:
size of the code, I think adding all the metrics in the main loop will be too much

Reply:
You could also define RATING_METRICS and RANKING_METRICS as dicts and then use for loops (as in the tuning notebooks). It would remove some stuff that repeats.

Reply:
this is how it is done, you can see this in cell 13:
```
rating_evaluator = {
    "als": lambda test, predictions: rating_metrics_pyspark(test, predictions),
    "svd": lambda test, predictions: rating_metrics_python(test, predictions),
    "fastai": lambda test, predictions: rating_metrics_python(test, predictions)
}
    
    
ranking_evaluator = {
    "als": lambda test, predictions, k: ranking_metrics_pyspark(test, predictions, k),
    "sar": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),
    "svd": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),
    "fastai": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),
    "ncf": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),
}
```
Le and Jun Ki were proposing to call the metrics directly, but I think it is too much code

Reply:
You could reduce some repetition here too using two dictionaries of the form
`{"als": rating_metrics_pyspark, "svd": rating_metrics_python, "fastai": rating_metrics_python}`
`{"als": ranking_metrics_pyspark, "sar": ranking_metrics_python, "svd": ranking_metrics_python, "fastai": ranking_metrics_python, "ncf": ranking_metrics_python}` etc.

Reply:
It is good to have these dictionaries anyway, because they can be reused by many of the notebooks we have written.

Reply:
hey @anargyri, I added a col dictionary, will be there in the next commit. I left the metrics as:
```
def rating_metrics_python(test, predictions):
    return {
        "RMSE": rmse(test, predictions, **COL_DICT),
        "MAE": mae(test, predictions, **COL_DICT),
        "R2": rsquared(test, predictions, **COL_DICT),
        "Explained Variance": exp_var(test, predictions, **COL_DICT)
    }
```
I think this is a good compromise between what Jun Ki and Le were thinking and having the metrics as a dict comprehension. We could argue that this is more explicit

Reply:
@miguelgfierro  Hey Miguel, sorry for un-resolving the conversation, but what do you think about adding functions "all_rating_metrics" and "all_ranking_metrics" to our evaluation module? All of our notebooks do that anyway. They will be the same functions as these that you have in benchmark_utils.

Reply:
I'm not super in favor tbh, I think it is the same reason that Le said, we are trying to do things explicitely over implicitely. In general situations like the notebooks, we have only 4 lines of code:
```
rmse(test, predictions, **cols),
mae(test, predictions, **cols),
rsquared(test, predictions, **cols),
exp_var(test, predictions, **cols)
```
This is not too much and I think it is nice to show it explicitly as opposite of having 1 line of implicit `rating_metrics`.

In the case of this notebook, where we have to add all the metrics at the same time, I think it could make sense to have it, just because if not it gets super messy. 

Does that make sense?

Reply:
Makes sense, I think it is flexible how much one may want to refactor a notebook, since it is partly development and partly demonstration.